libzfs/src/vdev.rs

```diff
@@ -0,0 +1,183 @@
+        whole_disk: Option<bool>,
+        is_log: Option<bool>,
+    },
+    File {
+        guid: Option<String>,
+        state: String,
+        path: String,
+        is_log: Option<bool>,
+    },
+}
+
+pub fn enumerate_vdev_tree(tree: &nvpair::NvList) -> Result<VDev> {
+    let tmp = tree.lookup_string(sys::zpool_config_type())?;
+    let x = tmp.as_bytes_with_nul();
+
+    fn get_children(tree: &nvpair::NvList) -> Result<Vec<VDev>> {
+        tree.lookup_nv_list_array(sys::zpool_config_children())?
+            .iter()
+            .map(enumerate_vdev_tree)
+            .collect()
+    }
+
+    fn get_spares(tree: &nvpair::NvList) -> Result<Vec<VDev>> {
+        let spares = tree.lookup_nv_list_array(sys::zpool_config_spares());
+
+        match spares {
+            Ok(x) => x.iter().map(enumerate_vdev_tree).collect(),
+            Err(_) => Ok(vec![]),
+        }
+    }
+
+    fn get_cache(tree: &nvpair::NvList) -> Result<Vec<VDev>> {
+        let cache = tree.lookup_nv_list_array(sys::zpool_config_l2cache());
+
+        match cache {
+            Ok(x) => x.iter().map(enumerate_vdev_tree).collect(),
+            Err(_) => Ok(vec![]),
+        }
+    }
+
+    fn lookup_tree_str(tree: &nvpair::NvList, name: String) -> Result<Option<String>> {
+        let x = tree.lookup_string(name);
+
+        match x {
+            Ok(x) => Ok(Some(x.into_string()?)),
+            Err(_) => Ok(None),
+        }
+    }
+
+    fn lookup_is_log(tree: &nvpair::NvList) -> Option<bool> {
+        tree.lookup_uint64(sys::zpool_config_is_log())
+            .map(|x| x == 1)
+            .ok()
+    }
+
+    fn lookup_guid(tree: &nvpair::NvList) -> Option<String> {
+        tree.lookup_uint64(sys::zpool_config_guid())
+            .map(|x| format!("{:#018X}", x))
+            .ok()
+    }
+
+    fn lookup_state(tree: &nvpair::NvList) -> Result<String> {
+        let vdev_stats = tree.lookup_uint64_array(sys::zpool_config_vdev_stats())
+            .map(sys::to_vdev_stat)?;
+
+        let state = unsafe {
+            let s = sys::zpool_state_to_name(
+                sys::to_vdev_state(vdev_stats.vs_state as u32).ok_or(Error::new(
+                    ErrorKind::NotFound,
+                    "vs_state not in enum range",
+                ))?,
+                sys::to_vdev_aux(vdev_stats.vs_aux as u32)
+                    .ok_or(Error::new(ErrorKind::NotFound, "vs_aux not in enum range"))?,
+            );
+
+            CStr::from_ptr(s)
+        };
+
+        state.to_owned().into_string().map_err(LibZfsError::from)
+    }
+
+    match x {
+        x if x == sys::VDEV_TYPE_DISK => {
+            let path = tree.lookup_string(sys::zpool_config_path())?.into_string()?;
+            let dev_id = lookup_tree_str(tree, sys::zpool_config_dev_id())?;
+            let phys_path = lookup_tree_str(tree, sys::zpool_config_phys_path())?;
+            let whole_disk = tree.lookup_uint64(sys::zpool_config_whole_disk())
+                .map(|x| x == 1)
```

Comment:
Does this represent that whole disk is true? If so, didn't you put in a patch for boolean types where 1 represents true? Should this be replaced with that type?

Reply:
> Does this represent that whole disk is true? 

Yes.

>  If so, didn't you put in a patch for boolean types where 1 represents true? 

No, the generated zfs bindings have a type where 0 == false & 1 == true.

> Should this be replaced with that type?

I don't think so. The ZFS source usages uses 0 & 1 for comparison with this result.